call_option.go

```diff
@@ -1,118 +1,97 @@
-// retrying.
-func WithTimeout(timeout time.Duration) CallOption {
-	return withTimeout(timeout)
-}
-
-type withRetryCodes []codes.Code
-
-func (w withRetryCodes) Resolve(s *CallSettings) {
-	s.RetrySettings.RetryCodes = make(map[codes.Code]bool)
-	for _, code := range []codes.Code(w) {
-		s.RetrySettings.RetryCodes[code] = true
+func (r *Backoff) pause() time.Duration {
+	if r.Initial == 0 {
+		r.Initial = time.Second
 	}
+	if r.Max == 0 {
+		r.Max = 30 * time.Second
+	}
+	if r.Mult == 0 {
+		r.Mult = 2
+	}
+	d := time.Duration(rand.Int63n(int64(r.Initial)))
+	r.Initial = time.Duration(float64(r.Initial) * r.Mult)
+	if r.Initial > r.Max {
+		r.Initial = r.Max
+	}
+	return d
 }
 
-// WithRetryCodes sets a list of Google API canonical error codes upon which a
-// retry should be attempted. If nil, the call will not retry.
-func WithRetryCodes(retryCodes []codes.Code) CallOption {
-	return withRetryCodes(retryCodes)
-}
-
-type withDelayTimeoutSettings MultipliableDuration
-
-func (w withDelayTimeoutSettings) Resolve(s *CallSettings) {
-	s.RetrySettings.BackoffSettings.DelayTimeoutSettings = MultipliableDuration(w)
-}
-
-// WithDelayTimeoutSettings specifies:
-// - The initial delay time, in milliseconds, between the completion of
-//   the first failed request and the initiation of the first retrying
-//   request.
-// - The multiplier by which to increase the delay time between the
-//   completion of failed requests, and the initiation of the subsequent
-//   retrying request.
-// - The maximum delay time, in milliseconds, between requests. When this
-//   value is reached, `RetryDelayMultiplier` will no longer be used to
-//   increase delay time.
-func WithDelayTimeoutSettings(initial time.Duration, max time.Duration, multiplier float64) CallOption {
-	return withDelayTimeoutSettings(MultipliableDuration{initial, max, multiplier})
-}
-
-type withRPCTimeoutSettings MultipliableDuration
-
-func (w withRPCTimeoutSettings) Resolve(s *CallSettings) {
-	s.RetrySettings.BackoffSettings.RPCTimeoutSettings = MultipliableDuration(w)
+type CallSettings struct {
+	// Retry returns a RetryFunc to be used to control retry logic of a method call.
+	// If Retry is nil or the returned RetryFunc is nil, the call will not be retried.
+	Retry func() RetryFunc
 }
 
-// WithRPCTimeoutSettings specifies:
-// - The initial timeout parameter to the request.
-// - The multiplier by which to increase the timeout parameter between
-//   failed requests.
-// - The maximum timeout parameter, in milliseconds, for a request. When
-//   this value is reached, `RPCTimeoutMultiplier` will no longer be used
-//   to increase the timeout.
-func WithRPCTimeoutSettings(initial time.Duration, max time.Duration, multiplier float64) CallOption {
-	return withRPCTimeoutSettings(MultipliableDuration{initial, max, multiplier})
-}
+// RetryFunc decides whether a request should be retried and how long to wait before retrying
+// if the previous attempt returned with err. err is never nil.
```

Comment:
I think "should" and "must" sounds like a constraint that implementer needs to uphold.
The intention of this comment, however, is to promise the implementer that Invoke will not call this function with a nil error. I've changed wording on this.